engine/consensus/compliance/engine.go

```diff
@@ -493,23 +515,27 @@ func (e *Engine) processBlockProposal(proposal *messages.BlockProposal) error {
 	// if the error is a known invalid extension of the protocol state, then
 	// the input is invalid
 	if state.IsInvalidExtensionError(err) {
+		stopMetrics()
 		return engine.NewInvalidInputErrorf("invalid extension of protocol state (block: %x, height: %d): %w",
 			header.ID(), header.Height, err)
 	}
 
 	// if the error is an known outdated extension of the protocol state, then
 	// the input is outdated
 	if state.IsOutdatedExtensionError(err) {
+		stopMetrics()
```

Comment:
This works, but the added metrics logic makes the function harder to read. 

Instead, we could do the following:

```
func (e *Engine) processBlockProposal(proposal) error {
  err := e.processCurrentProposalWithMetrics(proposal)
  if err ...

  err := e.processPendingChild(header)
  if err ...
}

func (e *Engine) processCurrentProposalWithMetrics(proposal) error {
  stopMetrics := ...
  defer stopMetrics()

  err := e.processCurrentProposal(proposal)
  if err ...
}

func (e *Engine) processCurrentProposal(proposal) error {
  ...
  // existing logic 
}
```

- all the metrics related code could go to `processCurrentProposalWithMetrics`
- it's clear now `processBlockProposal` will first process current proposal, then pending children